sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,492 @@
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public UnboundedMqttSource getCurrentSource() {
+      return source;
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return currentTimestamp;
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to write and send a message to a MQTT server.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<byte[]>, PDone> {
+
+    @Nullable abstract MqttConnectionConfiguration mqttConnectionConfiguration();
+    @Nullable abstract int qos();
+    @Nullable abstract boolean retained();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setMqttConnectionConfiguration(MqttConnectionConfiguration configuration);
+      abstract Builder setQos(int qos);
+      abstract Builder setRetained(boolean retained);
+      abstract Write build();
+    }
+
+    public Write withMqttConnectionConfiguration(MqttConnectionConfiguration configuration) {
+      return toBuilder().setMqttConnectionConfiguration(configuration).build();
+    }
+
+    public Write withQoS(int qos) {
+      return toBuilder().setQos(qos).build();
+    }
+
+    public Write withRetained(boolean retained) {
+      return toBuilder().setRetained(retained).build();
+    }
+
+    @Override
+    public PDone apply(PCollection<byte[]> input) {
+      input.apply(ParDo.of(new MqttWriter(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    @Override
+    public void validate(PCollection<byte[]> input) {
+      mqttConnectionConfiguration().validate();
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      mqttConnectionConfiguration().populateDisplayData(builder);
+      builder.add(DisplayData.item("qos", qos()));
+      builder.add(DisplayData.item("retained", retained()));
+    }
+
+    /**
+     * MQTT writer.
+     */
+    private static class MqttWriter extends DoFn<byte[], Void> {
+
+      Write spec;
+
+      private transient MqttClient client;
+
+      public MqttWriter(Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void createMqttClient() throws Exception {
+        client = spec.mqttConnectionConfiguration().getClient();
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        byte[] payload = context.element();
+        MqttMessage message = new MqttMessage();
+        message.setQos(spec.qos());
+        message.setRetained(spec.retained());
+        message.setPayload(payload);
+        client.publish(spec.mqttConnectionConfiguration().topic(), message);
+      }
+
+      @Teardown
+      public void closeMqttClient() throws Exception {
+        if (client != null) {
+          client.disconnect();
+          client.close();
```

Comment:
In the other place in code, you have this wrapped in a try/finally - is it necessary or unnecessary?